metrics/runtime_metrics.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// StartReportingRuntimeMetrics starts reporting runtime metrics under given metrics scope with
+// given frequency.
+// This is just a convenience wrapper around creating and starting a new reporter manually.
+// Recommended usage:
+//     StartReportingRuntimeMetrics(rootScope.Scope("runtime"), time.Second)
+// If you want to stop the reporter later:
+//     rmr := StartReportingRuntimeMetrics(rootScope.Scope("runtime"), time.Second)
+//     ...
+//     rmr.Close()
+func StartReportingRuntimeMetrics(
+	scope tally.Scope, reportInterval time.Duration,
+) *RuntimeReporter {
+	runtimeReporter := NewRuntimeReporter(scope, reportInterval)
+	runtimeReporter.Start()
+	return runtimeReporter
+}
+
+type runtimeMetrics struct {
+	numGoRoutines   tally.Gauge
+	goMaxProcs      tally.Gauge
+	memoryAllocated tally.Gauge
+	memoryHeap      tally.Gauge
+	memoryHeapIdle  tally.Gauge
+	memoryHeapInuse tally.Gauge
+	memoryStack     tally.Gauge
+	numGC           tally.Counter
+	gcPauseMs       tally.Timer
+	lastNumGC       uint32
+}
+
+// RuntimeReporter is a struct containing the state of the RuntimeMetricsReporter.
+type RuntimeReporter struct {
```

Comment:
this is going to get into hand-waving territory, but calling it a Reporter is very confusing, since it takes in a tally.Scope, which has a Reporter to send metrics over.

I'd prefer we call it something along the lines of "runtime metrics collector", since that's what it actually does.